### PR TITLE
gpt-generator, bootctl, nspawn: EFI mount point handling updates

### PR DESCRIPTION
**Pull request by poettering**
This updates bootctl quite a bit and makes it search for the EFI mount point at /boot, /efi and /boot/efi.

This updates nspawn to mount the ESP to /boot or /efi automatically if there's one in a disk image passed to --image=. This way "bootctl" can be used inside of nspawn containers.

This also updates the gpt generator to optionally use /efi as mount point if /boot is not available as mount point because something else is located there.

Finally, this updates kernel-install to install the kernel to /efi if /boot is blocked by something else.

### CODE REVIEW

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 1]
```diff
@@ -453,33 +453,37 @@ static int add_boot(const char *what) {
         _cleanup_blkid_free_probe_ blkid_probe b = NULL;
         const char *fstype = NULL, *uuid = NULL;
         sd_id128_t id, type_id;
+        const char *esp;
         int r;
 
         assert(what);
 
         if (!is_efi_boot()) {
```
Comment (keszybz) [inline src/gpt-auto-generator/gpt-auto-generator.c]: Hm, do we really want to skip mounting the ESP in this case? It seems reasonable to mount it, even if we boot from BIOS. Not related to this PR of course.

File: man/bootctl.xml [hunk 2]
```diff
@@ -102,7 +102,9 @@
       <xi:include href="standard-options.xml" xpointer="version" />
       <varlistentry>
         <term><option>--path</option></term>
-        <listitem><para>Path to the EFI system partition. The default is /boot.</para></listitem>
+        <listitem><para>Path to the EFI system partition (ESP). If not specified the ESP is searched at
```
Comment (keszybz) [inline man/bootctl.xml]: "searched at" sounds wrong. Maybe "If not specified, /efi, /boot, and /boot/efi are checked in turn."

File: src/gpt-auto-generator/gpt-auto-generator.c (hunk 1 above)
Comment (poettering) [inline src/gpt-auto-generator/gpt-auto-generator.c]: ah, indeed. you are right, we should just mount it always, if it is there. Will add another patch to this PR to drop this check.

File: man/bootctl.xml (hunk 2 above)
Comment (poettering) [inline man/bootctl.xml]: will fix

**Comment by poettering** [comment]
OK, I force pushed a new version that adds a commit that relaxes the EFI check in the generator a bit. I also rebased it on current git, and fixed the doc issue you pointed out. Please have a look and merge!

**Comment by fsateler** [comment]
> This updates bootctl quite a bit and makes it search for the EFI mount point at /boot, /efi and /boot/efi.

But the gpt-generator does not use /boot/efi. Is this intentional?

**Comment by poettering** [comment]
> This updates bootctl quite a bit and makes it search for the EFI mount point at /boot, /efi and /boot/efi.
> 
> But the gpt-generator does not use /boot/efi. Is this intentional?

Yes. Because /boot/efi is just stupid, as it implies that an ESP partition implies a separate /boot partition, but that's nonsense.

Also, a generator is run with nothing but the root dir plus /run mounted. That means even if we wanted we couldn't check if /boot/efi is a usable mount point as that as likely itself on a mount point that isn't mounted yet. Which is another reason why /boot/efi is stupid.

**Comment by martinpitt** [comment]
> Because /boot/efi is just stupid, as it implies that an ESP partition implies a separate /boot partition

No, it doesn't, why would it? The EFI partition must be separate of course, but distro installers don't configure a separate `/boot` partition on EFI (as that would indeed be silly).

**Comment by poettering** [comment]
> No, it doesn't, why would it? The EFI partition must be separate of course, but distro installers don't configure a separate /boot partition on EFI (as that would indeed be silly).

well, but /boot would effectively be empty then, only having a subdir? Why have it then?

**Comment by fsateler** [comment]
> > No, it doesn't, why would it? The EFI partition must be separate of course, but distro installers don't configure a separate /boot partition on EFI (as that would indeed be silly).
> 
> well, but /boot would effectively be empty then, only having a subdir? Why have it then?

Debian kernel packages ship the kernels in /boot:

```
$ dpkg -L linux-image-3.16.0-4-amd64|grep /boot
/boot
/boot/System.map-3.16.0-4-amd64
/boot/config-3.16.0-4-amd64
/boot/vmlinuz-3.16.0-4-amd64
```

Which is IMO nice to have on non-efi systems, otherwise you get to duplicate every kernel you have installed.

**Comment by poettering** [comment]
But /boot is a weird place for the kernels if they aren't needed there... On Fedora, kernels are shiped next to the kmods in /usr/lib/modules/$KERNEL_VERSION and simply copied into the ESP or /boot as necessary. That way the boot partition (regardless if ESP or legacy) is entirely redundant, and the system partition is fully self-sufficient, so that the ESP or /boot can be reconstructed at any time.

**Comment by keszybz** [comment]
Looks good to merge. Not merging because of the v232 milestone.

**Comment by jonathonf** [comment]
Doesn't this contradict the Linux FHS?

http://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s05.html

**Comment by keszybz** [comment]
FHS is mostly dead, it hasn't been updated to include most of the developments in the Linux ecosystem in the last decade. We don't want to deviate from FHS without reason, but for stuff that it simply doesn't cover because it wasn't around when it was written, we don't have much choice.

**Comment by nero-dv** [comment]
> FHS is mostly dead, it hasn't been updated to include most of the developments in the Linux ecosystem in the last decade. We don't want to deviate from FHS without reason, but for stuff that it simply doesn't cover because it wasn't around when it was written, we don't have much choice.

This is recursive logic. What's the reason that we deviated from the standard in this particular instance?

**Comment by fabiscafe** [comment]
@nero-dv There is no standard for modern (U)EFI-Boot in the FSH. That's something that is around for 11years on every new PC at this point (longer for specific cases). The FSH needs an update.

**Comment by mainettis** [comment]
This change is a pain and way too opinionated. It's making assumptions where things should be and hard coding them. This forced use of paths without checking preset values or any way to modify result in issues like #26644. And, several other work around and fixes since implementation.

As was stated before, it does violate [FHS3.0.03.05./boot](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s05.html) . And, no.. that FHS is not outdated, it simply states all boot required files go under root/boot folder. One should be able to mount any subdir in there ("/boot/efi", "/boot/esp", "/boot/esp(n)", etc..) Which is preferable to having multiple root directories filled with all sorts of boot only related files. There may be a use case requiring 5 partitions during boot. Should we then create 5 different directories in root?

Not only that, but it could possibly end up ignoring other [(U)EFI spec's additional device path protocol sub type file definitions](https://uefi.org/specs/UEFI/2.10/10_Protocols_Device_Path_Protocol.html). Of which there can be multiple. This change may also not be as distribution agnostic, since it doesn't seem to account for [registered subdirectories](https://uefi.org/registry) like "/EFI/Debian" during install.

I know this is old. And, I'm late to the comment. The damage is done and there have been lots of changes as a result. But, I don't re-install linux from scratch that often. Yet, this is the first time I've been stuck for days trying to get details on a new boot directory/partition "standard". There is just no discussion on this decision, nor reasoned argument for it.  And:
> Because /boot/efi is just stupid, as it implies that an ESP partition implies a separate /boot partition, but that's nonsense.

Is not a good argument. Since ESP partition implication would then mean that its proper mount point would be something like [/tmp/esp/](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s18.html). As it needs to be access by programs and install scripts. Therefore should not be mounted to [/mnt](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s12.html) as separate partitions usually are.

Nor is it:
> Also, a generator is run with nothing but the root dir plus /run mounted. That means even if we wanted we couldn't check if /boot/efi is a usable mount point as that as likely itself on a mount point that isn't mounted yet. Which is another reason why /boot/efi is stupid.

It is an overthought. Since it could be true of "/efi" path. As technically we could have a separate esp trying to mount to the same location. Same as multiple partitions trying to mount to "/boot" for some reason. But, chances are that most are going to mount and mkdir the parent before the child. And, will deal with consequences of mistakes outside the tool's scope. The same way we setup up and mount root before "/efi" or fail. Much simpler to attempt default or return an error when boot path structures are not defined by conf or other parameters.

**Comment by gregory-lee-bartholomew** [comment]
A single $BOOT partition is the way to go.

https://github.com/lnussel/lnussel.github.io/blob/fs/_posts/2020-12-16-fslayout.md#boot-files

**Pull request merged**